mesonbuild/interpreter/compiler.py

```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
     @noKwargs
     def get_argument_syntax_method(self, args: T.List['TYPE_var'], kwargs: 'TYPE_kwargs') -> str:
         return self.compiler.get_argument_syntax()
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+        KwargInfo('api', (str, NoneType)),
+        KwargInfo('compilation', (str, NoneType)),
+        KwargInfo('static_compilation', (str, NoneType)),
+        KwargInfo('static_only', bool, default=False),
+        )
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '0.63.0')
+    def symbol_visibility_header_method(self, args: T.List[str], kwargs: 'TYPE_kwargs') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = ''
+
+        if static_only:
+            content += textwrap.dedent(f'''\
+                #ifndef {static_compilation}
+                #  define {static_compilation}
+                #endif /* {static_compilation} */
+                ''')
+
+        if self.compiler.info.is_windows():
```

Comment:
This is something I found really annoying about the cmake version, because you generate a file that isn't comprehensibly useful in the general case. What's the point in saving a few bytes for cross-platform defines?

Reply:
I'm not sure it is possible to have a static file that works in all cases tbh. How do you detect you're building a DLL, regardless of the compiler, from within a macro?

Reply:
This review comment is responding to the fact that the file contents are generated by checking in meson code whether you're building for Windows or `GNUC`, instead of generating both everywhere and using preprocessor checks to determine which one is used.

Generating both everywhere and using preprocessor checks to determine which one is used, results in a header that serves as a good teaching example and can be extracted from any OS and embedded into a third-party derivation.

I have been previously annoyed that cmake generates Windows-specific snippets, which means that the example file it generates -- and which people try to use in PRs to the WrapDB -- cannot be reused in a Meson port.

Reply:
Ah, I see, makes more sense, I'll update.

Reply:
I changed the code to have a common block for all platforms, but Meson still have to define `MESON_HOST_IS_WINDOWS` or `MESON_HAS_GNU_SYMBOL_VISIBILITY`. I don't think the latter can be determined by preprocessor, can it? The former could be replaced by something like `#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__NT__)` but I'm actually not entirely sure what's the right case where we actually need dllexport/dllimport, clang-cl, cygwin?

Reply:
What, actually there is an example there: https://gcc.gnu.org/wiki/Visibility.

Reply:
Ok updated again, now without any check at meson level. Is that what you had in mind?